sdk/src/extras/disk.rs

```diff
@@ -331,118 +331,92 @@ pub mod types {
-                    let encoded = base64::engine::general_purpose::STANDARD.encode(&chunk[0..n]);
-
-                    if senders[i % self.upload_thread_ct]
-                        .send((offset, encoded, n as u64))
-                        .await
-                        .is_err()
-                    {
-                        // Failure to send indicates that the upload task exited early
-                        // due to an error on its end. We will return that error below.
-                        break Ok(());
-                    }
-                } else {
-                    // Bump the progress bar here to make it consistent
-                    self.upload_progress.fetch_add(n as u64, Ordering::Relaxed);
-                    self.progress_tx
-                        .send_replace(self.upload_progress.load(Ordering::Relaxed));
+            // Resubmit any failed offsets back to the remaining workers to retry.
+            while let Ok(failed_offset) = failed_rx.recv_async().await {
+                if resubmit_tx.send_async(failed_offset).await.is_err() {
+                    // All worker tasks have failed.
+                    break;
                 }
-
-                offset += CHUNK_SIZE;
-                i += 1;
-            };
-
-            for tx in senders {
-                drop(tx);
             }
+            drop(resubmit_tx);
 
             let mut errors = Vec::new();
-            if let Err(e) = read_result {
-                errors.push(e);
-            }
-
             for handle in handles {
                 let result = handle.await.map_err(DiskImportError::other)?;
                 if let Err(err) = result {
                     errors.push(err);
                 }
             }
 
-            match errors.len() {
-                1 => {
-                    return Err(DiskImportError::context(
-                        "Error while uploading the disk image",
-                        errors.remove(0),
-                    ))
-                }
-                2.. => {
-                    let mut msg = String::from("Errors while uploading the disk image:");
-                    for err in errors {
-                        msg += &format!("\n * {err}");
+            // Only return an error if all worker tasks failed.
+            if errors.len() == self.upload_thread_ct {
+                match errors.len() {
+                    1 => {
+                        return Err(DiskImportError::context(
+                            "Error while uploading the disk image",
+                            errors.remove(0),
+                        ))
+                    }
+                    2.. => {
+                        let mut msg = String::from("Errors while uploading the disk image:");
+                        for err in errors {
+                            msg += &format!("\n * {err}");
+                        }
+                        return Err(DiskImportError::Other(msg.into()));
```

Comment:
this no longer seems right: the default use will produce 10 errors, mostly the same--is that correct?

Reply:
Updated to dedupe the error messages shown to the user.